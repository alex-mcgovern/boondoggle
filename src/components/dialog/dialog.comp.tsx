import { Slot } from "@radix-ui/react-slot";
import clsx from "clsx";
import { createRef, forwardRef, useCallback } from "react";

import { useClickOutside } from "../../hooks/use_click_outside";
import { useForwardRef } from "../../hooks/use_forward_ref";
import { useOpenDialogWithKeyboard } from "../../hooks/use_open_dialog_with_keyboard";
import { a11yFocus } from "../../styles/common/a11y.css";
import { animateAppear } from "../../styles/common/animations.css";
import { Box } from "../box";
import { getDialogStyles } from "./dialog.styles.css";

import type { BoxProps } from "../box";
import type { DialogPlacementEnum } from "./dialog.styles.css";
import type { ReactNode } from "react";

export type DialogProps = BoxProps & {
  children?: ReactNode;
  className?: string;
  inert?: boolean;
  isOpen?: boolean;
  onIsOpenChange?: (isOpen?: boolean) => void;
  placement?: DialogPlacementEnum;
  preventOpenOnKeydown?: boolean;
  triggerNode?: ReactNode;
  wrapperProps?: BoxProps;
};

export const Dialog = forwardRef<HTMLDialogElement, DialogProps>(
  (
    {
      children,
      className: userClassName,
      isOpen: controlledIsOpen,
      onIsOpenChange,
      preventOpenOnKeydown = false,
      triggerNode,
      placement,
      wrapperProps,
      ...rest
    }: DialogProps,
    ref
  ) => {
    const dialogRef = useForwardRef<HTMLDialogElement>(ref);

    const triggerRef = createRef<HTMLElement>();

    /**
     * Callback for when the input is clicked or focused.
     */
    const toggleIsOpen = useCallback(() => {
      if (!dialogRef.current) {
        return undefined;
      }

      return (
        dialogRef.current?.open
          ? dialogRef.current?.close()
          : dialogRef.current?.show(),
        onIsOpenChange?.(dialogRef.current?.open)
      );
    }, [dialogRef, onIsOpenChange]);

    /**
     * When input is focused and user presses enter, open the date picker.
     */
    useOpenDialogWithKeyboard<HTMLElement>({
      callback: toggleIsOpen,
      dialogRef,
      preventOpenOnKeydown,
      triggerRef,
    });

    /**
     * Handle click outside dialog
     */
    useClickOutside<HTMLDialogElement, HTMLElement>({
      callback: () => {
        return dialogRef.current?.close();
      },
      contentRef: dialogRef,
      triggerRef,
    });

    return (
      <Box position="relative" {...wrapperProps}>
        <Slot onClick={toggleIsOpen} ref={triggerRef}>
          {triggerNode}
        </Slot>
        <Box
          {...rest}
          as="dialog"
          open={controlledIsOpen}
          ref={dialogRef}
          className={clsx(
            animateAppear,
            userClassName,
            a11yFocus,
            getDialogStyles({ placement })
          )}
        >
          {children}
        </Box>
      </Box>
    );
  }
);
