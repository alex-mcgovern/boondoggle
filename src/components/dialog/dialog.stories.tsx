import React from "react";

import { Dialog as StoryComponent } from ".";
import { WithDarkLightMode } from "../../../.storybook/with_dark_light_mode";
import { LOREM } from "../../../mocks/LOREM.mock";
import { Button } from "../button";

import type { DialogProps as StoryComponentProps } from ".";
import type { StoryFn, StoryObj } from "@storybook/react";

export default {
  title: "React components/Dialog",
  component: StoryComponent,
};

const Template: StoryFn<StoryComponentProps> = ({
  ...rest
}: StoryComponentProps) => {
  return <StoryComponent {...rest} />;
};

export const Default: StoryObj<StoryComponentProps> = {
  render: Template,
  name: "default",
  parameters: {
    layout: "fullscreen",
  },
  decorators: [WithDarkLightMode],
  args: {
    triggerNode: <Button name="Dialog button">Open dialog</Button>,
    children: LOREM.text_xxl,
    title: LOREM.text_xxs,
  },
};
