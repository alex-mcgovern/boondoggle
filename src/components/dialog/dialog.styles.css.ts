import { keyframes, style } from "@vanilla-extract/css";

import { vars } from "../../styles/theme.css";
import { createAccessibleTransition } from "../../styles/utils/create_accessible_transition";
import { getSprinkles } from "../../styles/utils/get_sprinkles.css";

const overlayShow = keyframes({
  "0%": { opacity: 0 },
  "100%": { opacity: 1 },
});

const contentShow = keyframes({
  "0%": { opacity: 0, transform: "translate(-50%, -48%) scale(.96)" },
  "100%": { opacity: 1, transform: "translate(-50%, -50%) scale(1)" },
});

export const dialogOverlay = style([
  {
    background: `#00000080`,
    position: "fixed",
    inset: 0,
    zIndex: 99999,
  },
  createAccessibleTransition({
    animation: `${overlayShow} ${vars.transitionDuration.medium} cubic-bezier(0.16, 1, 0.3, 1) forwards`,
  }),
]);

export const dialogContent = style([
  getSprinkles({
    display: "block",
    isolation: "isolate",
    overflow: "hidden",
    textDecoration: "none",
    background: "tint_default",
    border: "border_default",
    borderRadius: "md",
    boxShadow: "lg",
    width: {
      mobile: "gridSpan10",
      tablet: "gridSpan8",
      desktop: "gridSpan8",
    },
    maxWidth: "gridSpan10",
  }),
  {
    zIndex: 99999,
    position: "fixed",
    top: "50%",
    left: "50%",
    transform: "translate(-50%, -50%)",
    selectors: {
      "&:focus": { outline: "none" },
    },
  },
  createAccessibleTransition({
    animation: `${contentShow} ${vars.transitionDuration.long} ease forwards`,
  }),
]);

export const dialogInner = style([
  getSprinkles({
    maxHeight: "90vh",
    padding: "spacing2",
  }),
  {
    overflowY: "auto",
  },
]);
