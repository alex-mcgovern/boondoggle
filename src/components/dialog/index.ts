export { Dialog } from "./dialog.component";
export type { DialogProps } from "./dialog.component";
