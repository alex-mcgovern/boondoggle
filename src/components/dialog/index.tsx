import * as RadixSlot from "@radix-ui/react-slot";
import clsx from "clsx";
import { createRef, forwardRef, useCallback } from "react";

import { useClickOutside } from "../../hooks/use-click-outside";
import { useOpenDialogWithKeyboard } from "../../hooks/use-open-dialog-with-keyboard";
import { useForwardRef } from "../../hooks/use-forward-ref";
import { Box } from "../box";
import { getDialogStyles } from "./styles.css";

import type { ReactNode } from "react";
import type { BoxProps } from "../box";
import type { DialogPlacementEnum } from "./styles.css";

export type DialogProps = BoxProps & {
	/**
	 * Children to render inside the dialog component.
	 */
	children?: ReactNode;

	/**
	 * Optional CSS class name applied to the outer HTML dialog element.
	 */
	className?: string;

	/**
	 * Whether the dialog is "inert" (can't be focused, can't be clicked.)
	 */
	inert?: boolean;

	/**
	 * Controls the dialog's open state, making it a controlled element. Use alongside `onIsOpenChange`.
	 */
	isOpen?: boolean;

	/**
	 * Callback when the dialog's open state is changed. Use alongside `isOpen`.
	 */
	onIsOpenChange?: (isOpen?: boolean) => void;

	/**
	 * Controls the interaction mode for the dialog trigger node.
	 */
	openOn?: "click" | "hover";

	/**
	 * Controls the dialog's placement relative to the trigger node.
	 */
	placement?: DialogPlacementEnum;

	/**
	 * Whether to open the dialog on an enter keypress while the trigger node is focused.
	 */
	preventOpenOnKeydown?: boolean;

	/**
	 * The react node to act as the trigger for the dialog.
	 */
	triggerNode?: ReactNode;

	/**
	 * `BoxProps` to pass to the outer element that wraps the dialog and it's trigger.
	 */
	wrapperProps?: BoxProps;
};

export const Dialog = forwardRef<HTMLDialogElement, DialogProps>(
	(
		{
			children,
			className: userClassName,
			isOpen: controlledIsOpen,
			onIsOpenChange,
			openOn = "click",
			placement,
			preventOpenOnKeydown = false,
			triggerNode,
			wrapperProps,
			...rest
		}: DialogProps,
		ref,
	) => {
		const dialogRef = useForwardRef<HTMLDialogElement>(ref);

		const triggerRef = createRef<HTMLElement>();

		const toggleIsOpen = useCallback(() => {
			if (!dialogRef.current) {
				return;
			}

			if (dialogRef.current?.open) {
				dialogRef.current?.close();
			}

			if (!dialogRef.current?.open) {
				dialogRef.current?.show();
			}

			onIsOpenChange?.(dialogRef.current?.open);
		}, [dialogRef, onIsOpenChange]);

		useOpenDialogWithKeyboard<HTMLElement>({
			callback: toggleIsOpen,
			dialogRef,
			preventOpenOnKeydown,
			triggerRef,
		});

		useClickOutside<HTMLDialogElement, HTMLElement>({
			callback: () => {
				return dialogRef.current?.close();
			},
			contentRef: dialogRef,
			triggerRef,
		});

		return (
			<Box position="relative" {...wrapperProps}>
				<Box background="transparent">
					<RadixSlot.Slot
						onClick={toggleIsOpen}
						onMouseEnter={
							openOn === "hover"
								? () => {
										return dialogRef.current?.show();
								  }
								: undefined
						}
						onMouseLeave={
							openOn === "hover"
								? () => {
										return dialogRef.current?.close();
								  }
								: undefined
						}
						onPointerEnter={
							openOn === "hover"
								? () => {
										return dialogRef.current?.show();
								  }
								: undefined
						}
						onPointerLeave={
							openOn === "hover"
								? () => {
										return dialogRef.current?.show();
								  }
								: undefined
						}
						onPointerOver={
							openOn === "hover"
								? () => {
										return dialogRef.current?.show();
								  }
								: undefined
						}
						ref={triggerRef}
					>
						{triggerNode}
					</RadixSlot.Slot>
				</Box>
				<Box
					{...rest}
					as="dialog"
					className={clsx(
						userClassName,
						getDialogStyles({ placement }),
					)}
					open={controlledIsOpen}
					ref={dialogRef}
				>
					<Box>{children}</Box>
				</Box>
			</Box>
		);
	},
);
