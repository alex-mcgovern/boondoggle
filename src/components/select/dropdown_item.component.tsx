import clsx from "clsx";
import React, { forwardRef } from "react";

import { Box } from "../box";
import * as styles from "./select.styles.css";

import type { SharedUiScale } from "../..";
import type { DropdownItemShape } from "./select.types";
import type { Ref } from "react";

export interface DropdownItemProps {
  isHighlighted: boolean;
  isMultipleSelectionEnabled?: boolean;
  size?: SharedUiScale;
  isMulti?: boolean;
  item: DropdownItemShape;
  isDropdownItemSelected?: boolean;
}

export const DropdownItem = forwardRef(
  (
    {
      item,
      size,
      isHighlighted,
      isMulti,
      isDropdownItemSelected,
      ...rest
    }: DropdownItemProps,
    ref
  ) => {
    return (
      <Box
        as="button"
        id={item.label}
        className={clsx(
          styles.getDropdownItemStyles({
            size,
          }),
          {
            [styles.isHighlighted]: isHighlighted,
            [styles.isSelected]: isDropdownItemSelected,
          }
        )}
        {...rest}
        ref={ref as Ref<HTMLButtonElement>}
      >
        <Box flexShrink="0">{item.label}</Box>
        {isMulti && (
          <Box
            as="input"
            type="checkbox"
            readOnly
            tabIndex={-1}
            checked={isDropdownItemSelected}
            marginLeft="auto"
          />
        )}
      </Box>
    );
  }
);
