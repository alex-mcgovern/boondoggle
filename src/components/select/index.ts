export { SelectSingle } from "./select_single.comp";
export type { SelectSingleProps } from "./select_single.comp";

export { SelectButton } from "./select_button.comp";
export type { SelectButtonProps } from "./select_button.comp";

export { SelectMulti } from "./select_multi.comp";
export type { SelectMultiProps } from "./select_multi.comp";

export type { DropdownItemShape } from "./select.types";
