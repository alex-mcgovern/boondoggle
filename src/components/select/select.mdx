import { ArgTypes, Canvas, Meta, Story } from "@storybook/addon-docs";

import * as stories from "./select.stories.tsx";

<Meta of={stories} />

# Select

- Wrapper around Downshift to provide a select component with a few options for
  styling and behavior

## Select Single

<Canvas name="default" of={stories.SelectSingleDefault} />

## Select multiple

<Canvas name="default" of={stories.SelectMultipleDefault} />

## Filterable

[ToDo: Add filterable example]
{/* <Canvas name="filterable" of={stories.Filterable} /> */}
