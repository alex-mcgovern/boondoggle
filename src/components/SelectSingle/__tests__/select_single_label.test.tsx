/** @jest-environment jsdom */
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { SelectSingle } from "..";
import { LOREM } from "../../../../mocks/LOREM.mock";
import "../../../../test/mocked_dependencies/dialog.mock";
import { mockSelectItems } from "../../Select/__mocks__/select.mock";

import type { SelectSingleProps } from "..";

/** ----------------------------------------------------------------------------- */

const PROPS: SelectSingleProps = {
  id: LOREM.id(),
  items: mockSelectItems({}),
  label: LOREM.label(),
  name: LOREM.text_xxs,
  placeholder: LOREM.select,
};

const renderComponent = (props: SelectSingleProps) => {
  return {
    user: userEvent.setup(),
    ...render(<SelectSingle {...props} />),
  };
};

/** ----------------------------------------------------------------------------- */

describe("<SelectSingle />", () => {
  it("should be labelled", async () => {
    const { container, getByRole } = await renderComponent(PROPS);

    const combobox = getByRole("combobox");
    const label = container.querySelector("label");
    const labelId = label?.getAttribute("id");

    expect(combobox.getAttribute("aria-labelledby")).toBe(labelId);
    expect(label).not.toBeNull();
    expect(label?.getAttribute("for")).toBe(PROPS.id);
    expect(label?.textContent).toBe(PROPS.label);
  });

  it("should match snapshot", async () => {
    const { getByRole } = await renderComponent(PROPS);

    expect(getByRole("combobox")).toMatchSnapshot();
  });
});
