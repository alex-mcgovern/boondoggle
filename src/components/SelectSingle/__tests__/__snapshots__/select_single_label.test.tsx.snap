// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectSingle /> should match snapshot 1`] = `
<input
  aria-activedescendant=""
  aria-autocomplete="list"
  aria-controls="downshift-2-menu"
  aria-expanded="false"
  aria-labelledby="downshift-2-label"
  autocomplete="off"
  class="styles_offsetSlotRightStyle__crhmtdd styles_getInputStyles__qrymjv9 styles__qrymjv8 get_sprinkles_width_100%_mobile__eiij7bgl a11y_a11yDisabled__r8he5t1 styles_getInputStyles_hasBorder_true__qrymjvb styles_hasBorder_true__qrymjv7 styles_getInputStyles_size_md__qrymjvd styles_size_md__qrymjv4 element_size_elementFontSize_md__1f6zrv4 element_size__1f6zrv1 get_sprinkles_fontSize_body_md__eiij7b21 get_sprinkles_lineHeight_body_md__eiij7b3f element_size_elementPadding_md__1f6zrva element_size__1f6zrv7 get_sprinkles_paddingLeft_spacing_1.5_mobile__eiij7bdx get_sprinkles_paddingRight_spacing_1.5_mobile__eiij7beu get_sprinkles_paddingTop_spacing_1_mobile__eiij7bfo get_sprinkles_paddingBottom_spacing_1_mobile__eiij7bcx styles__qrymjv1 get_sprinkles_height_spacing_5__eiij7b33 select_input.styles_selectInputCursorStyles__vh0x6y0"
  id="lorem-ipsum"
  name="Lorem ipsum"
  placeholder="Select a thing"
  readonly=""
  role="combobox"
  value=""
/>
`;
